src/tunnel/transport/io.rs

```diff
@@ -65,6 +65,12 @@ pub async fn propagate_local_to_remote(
             warn!("error while writing to tx tunnel {}", err);
             break;
         }
+
+        // Handle outstanding messages
+        if let Err(err) = ws_tx.handle_message().await {
```

Comment:
Handle message is never invoked/called if there is nothing to write on the websocket.
So there can be an unbounded amount of time between when the call is called.

You should move it into the select! and await for it. To be able to send frame on demand/when needed.

Reply:
The problem is that both `buf_mut` and `handle_message` need to borrow `ws_tx` exclusively, so we cannot call them concurrently in the select statement.
We know that it's actually OK, since they will mutate different fields that don't interact, but I don't know how to convince rustc of this without needing to leak these details into the `TunnelWrite` trait.

Reply:
One option could be to have another timer, that fires more frequently than the ping one, and return an option from the select statement. The option is then either a some with the length that was read from the local buffer, or a None indicating that one of the timers won the race in the select!.
We can then handle the ws write when we got a some, and call handle message in both cases.

Reply:
Ok, I refactored things a bit so that we can `select!` on two futures, that are waiting on data from the tunnel and messages from the reader.
I had to put this in a second `select!` inside the websocket writer to avoid issues with multiple exclusive borrows.

Reply:
One thing you can do is turn `ping_state: PingState` from WebsocketTunnelWrite into `ping_state: Option<PingState>` so you can use https://doc.rust-lang.org/std/option/enum.Option.html#method.take to steal/extract the ping state from the struct. With this ping state will be standalone and you will be free to mutate it separatly.

Reply:
It's not just the ping state though, we also need to read from the channel between the halves and send out frames.

And since we need our futures to be `Send`, that would require using `Mutex` or `RwLock` instead of `RefCell`, wouldn't it? So then we're back to locking in the read/write loop.

I realised that the current implementation is also not cancel-safe though, so it still needs some work in any case.

I'm thinking now to not expose the buffer as a trait method, but instead have a `recv` method on the `TunnelWrite` trait, that takes in `local_rx` and does a `select!` on both the `local_rx` and the message channel. This is then cancel-safe and we can run that in the `select!` statement. If this is the future that wins the select, we then call a write method that checks which buffer contains data (the one with the WS data or the one with the messages) and handles the data that it finds.

So this solves the issue with multiple borrows, as we'll do both receive/read operations inside the impl and not in `transport/io.rs`. It is cancel safe, as we only receive in the `select!` statement and handle the data outside. And it avoids poluting the `TunnelWrite` trait with websocket-specific details.

I'll try to implement this in the next couple of days when I find the time.